arrow-docs/docs/fx/managing-state/README.md

```diff
@@ -0,0 +1,188 @@
+
+The problem of having shared state throughout an application is consistency when this state is read and updated.
+An invariant checked a few lines ago may now no longer be valid if other threads changed the variables that were read.
+Updating a variable based on a prior read may lead to inconsistent state if there has been another update in-between...
+
+These issues are referred to as race-conditions and are the source of many headaches and lots of time spent debugging.
+
+In this article we will explore how arrow fx helps to deal with this issue and in some cases can even avoid it entirely.
+
+## Accessing a single shared variable
+
+When we are accessing only a single variable we can use [Atomic](TODO Link).
+This gives us access to a single mutable variable where reads and writes are atomic operations,
+ meaning no other access to the variable can happen during any operation on an atomic variable.
+ 
+```kotlin:ank:playground
+import arrow.fx.coroutines.*
+
+suspend fun main() {
+  val count = Atomic(0)
+  (0 until 20_000).parTraverse {
+    count.update(Int::inc)
+  }
+  println(count.get())
+}
+```
+
+In this example `count.update` is our atomic operation, and it is performed by possibly many threads at the same time.
+Despite that our counter increments one by one and never contains an invalid state (like say two threads incrementing 1 to 2 at the same time and both persisting 2 missing one update).
+
+### Drawbacks of using atomic variables
+
+Atomic variables have one major drawback: Their atomic guarantees do not scale beyond one variable.
+We can guarantee that access to one variable is atomic, but accessing two variables with atomic operations, even right after each other, will not make the entire operation atomic.
+
+## Securing access to shared state with locks
+
+One solution to the problem of modifying multiple variables of shared state is to surround the access with taking and releasing a lock.
+The general idea is to section off a piece of code that only one single thread can access at a time.
+
+In arrow fx we offer [ConcurrentVar](TODO link) as a datatype to build locking code with.
+A `ConcurrentVar` can either be empty or full and its operations `read` and `write` operations (called `take` and `put`) block if the variable is empty, for reads, or full, for writes.
+
+With a `ConcurrentVar` our example becomes:
+```kotlin:ank:playground
+import arrow.fx.coroutines.*
+
+suspend fun main(): Unit {
+  val mvar = ConcurrentVar(0)
+  (0 until 20_000).parTraverse {
+    val current = mvar.take()
+    // <1>
+    mvar.put(current + 1)
+  }
+  println(mvar.take())
+}
+```
+
+When a thread tries to invoke `ConcurrentVar.take()` it will block until the `ConcurrentVar` has a value to take and then removes the value.
+This means, after taking, no other thread can finish a call to `ConcurrentVar.take()` until `ConcurrentVar.put()` is called putting a value in the `ConcurrentVar`.
```

Comment:
This is not correct. Another thread can finish a call to `ConcurrentVar.take()`, if a concurrent thread in the meanwhile calls `put`. This doesn't happen in the above example, but it's still possible.

Reply:
Isn't that a terrible usage pattern? I think it is bad practice to `put` to a `ConcurrentVar` from a different section of code that has called `take` on it. I mean there is a good reason the most common use in the haskell world is `withMVar` which does all the taking and releasing in a resource safe manner.

Either way I'll try to adjust the wording a bit^^